harness/determined/common/api/_session.py

```diff
@@ -113,7 +112,59 @@ def _do_request(
         timeout: Optional[int],
         stream: bool,
     ) -> requests.Response:
-        pass
+        # Allow the json to come pre-encoded, if we need custom encoding.
+        if json is not None and data is not None:
+            raise ValueError("json and data must not be provided together")
+
+        if json:
+            data = det.util.json_encode(json)
+
+        relpath = path.lstrip("/")
+
+        # Use persistent session or create a new single-use one.
+        session = self._http_session or self._make_http_session()
+        try:
+            r = session.request(
+                method=method,
+                url=f"{self.master}/{relpath}",
+                params=params,
+                data=data,
+                headers=headers,
+                stream=stream,
+                timeout=timeout,
+            )
+        except requests.exceptions.SSLError:
+            raise
+        except requests.exceptions.ConnectionError as e:
+            raise errors.MasterNotFoundException(str(e))
+        except requests.exceptions.RequestException as e:
+            raise errors.BadRequestException(str(e))
+        finally:
+            if not self._http_session:
+                # Close the session if not persistent.
+                session.close()
+
+        def _get_error_str(r: requests.models.Response) -> str:
```

Comment:
This already violated our style guide (https://google.github.io/styleguide/pyguide.html#264-decision), and I think it's worse now that it's in a class.

> Avoid nested functions or classes except when closing over a local value other than self or cls. Do not nest a function just to hide it from users of a module. Instead, prefix its name with an _ at the module level so that it can still be accessed by tests.

Not your job, but bonus if you'd like to un-nest it.

Reply:
Thanks!